wordpress-mobile/WordPress-Android
# Integrate encrypted logging

This PR integrates encrypted logging from FluxC. The implementation on the WPAndroid side is very simple so it should be easy to review the code. However, testing these changes properly is very important since we can't cover all our basis on FluxC side.


**Known issues** (copied from https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1650)
1. Even though [we check for the file's existence](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/feature/encrypted-logs-polish/fluxc/src/main/java/org/wordpress/android/fluxc/store/EncryptedLogStore.kt#L127), I sometimes have a crash in a real device for the missing file [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/feature/encrypted-logs-polish/fluxc/src/main/java/org/wordpress/android/fluxc/store/EncryptedLogStore.kt#L140): https://sentry.io/share/issue/5be303a9982544ce852cd63ae2f5293f/. This hasn't happened on an emulator. I don't know how this is even possible with the `exists()` check before reading the contents. I'd have to look into this one, but want to get the other improvements out of the way first.
2. Log files randomly go missing on device if we can't upload them immediately. This feels like an issue with our log file handling in WPAndroid, so I'll look into that.
3. The uploaded log files still show up as a single line in MC. I had a very poor attempt at fixing it in this PR [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/feature/encrypted-logs-polish/fluxc/src/main/java/org/wordpress/android/fluxc/model/encryptedlogging/LogEncrypter.kt#L21) but I haven't yet prioritized looking into it.


**To test:**

0. Apply [this diff](https://cloudup.com/c5ijH7WXAFW) after which crashes will be reported in `DEBUG`. Also, tapping on the "Site Pages" and "Blog Posts" will immediately crash.
1. Clean install
2. Trigger a crash
3. Re-open the app
4. Wait for a bit (could take a few minutes)
5. Find the crash in sentry and use the `uuid` field (almost at the bottom) to check if the logs are available in MC
6. If the logs are not available yet, close the app and re-launch it and go back to step 4
7. Compare the logs from Sentry and MC. MC one will likely have 2 extra lines, but otherwise they should be the same.

---

0. Apply [this diff](https://cloudup.com/c5ijH7WXAFW) after which crashes will be reported in `DEBUG`. Also, tapping on the "Site Pages" and "Blog Posts" will immediately crash.
1. Clean install
2. Enable airplane mode (turn off internet connectivity)
3. Trigger a crash
4. Re-open the app
5. Open `Stetho -> Web SQL -> wp-fluxc -> EncryptedLogModel` table and verify there is one record
6. Trigger a different crash
7. Re-open the app
8. Open `Stetho -> Web SQL -> wp-fluxc -> EncryptedLogModel` table and verify there are two records
9. Close the app, turn off airplane mode, re-open the app (It's important that there is internet connectivity at application launch because, right now, that's the only case where we upload the encrypted logs)
11. Wait for a bit (could take a few minutes)
10. Find the crash in sentry and use the `uuid` field (almost at the bottom) to check if the logs are available in MC
11. If the logs are not available yet, close the app and re-launch it and go back to `step 9`
12. Compare the logs from Sentry and MC. MC one will likely have 2 extra lines, but otherwise they should be the same.

---

**PR submission checklist:**


_Leaving these ^ unchecked for now so I can look into them before taking it out of draft state._
